hta/common/trace_stack_filter.py

```diff
@@ -0,0 +1,294 @@
+        position: The position where the operator occurs in the trace. 0 for first occurrence, -1
+            for last occurrence.
+        method: How to filter out the events. See the enumerator class `OperatorFilterMethod`.
+        include_gpu_kernels: Whether to include the GPU kernel before the selected events.
+        name_column: Optional; name of the data frame column containing the operator name.
+            Default: "" (check DataFrame to use the available one).
+
+    Notes:
+        + When the op_name is not found in the DataFrame, the filter will return an empty DataFrame.
+    """
+
+    def __init__(
+        self,
+        op_name: str,
+        position: int,
+        method: OperatorFilterMethod,
+        include_gpu_kernels: bool = False,
+        name_column: Optional[str] = None,
+    ) -> None:
+        self.op_name = op_name
+        self.position = position
+        self.method = method
+        self.include_gpu_kernels = include_gpu_kernels
+        self.name_column: str = name_column if name_column else ""
+
+    def __call__(
+        self, df: pd.DataFrame, symbol_table: Optional[TraceSymbolTable] = None
+    ) -> pd.DataFrame:
+        if self.name_column == "":
+            name_column, cat_column = get_symbol_column_names(df)
+        else:
+            name_column = self.name_column
+            cat_column = self.name_column.replace("name", "cat")
+
+        if (
+            name_column not in df.columns
+            or cat_column not in df.columns
+            or df.dtypes[name_column] != "object"
+            or df.dtypes[cat_column] != "object"
+        ):
+            logger.error(f"df has no string column {name_column} of {cat_column}")
+            return df
+
+        found, op = find_op_occurrence(
+            df, self.op_name, position=self.position, name_column=name_column
+        )
+        if not found:
+            logger.warning(f"op {self.op_name}/{self.position} doesn't exist.")
+            return pd.DataFrame()
+
+        if self.method.value == OperatorFilterMethod.Under.value:
+            latest_start, earliest_end, stream = (
+                max(op["ts"], df["ts"].min()),
+                min(op["end"], df["end"].max()),
+                op["stream"],
+            )
+            df_ops = df.loc[
+                (df["ts"].ge(latest_start))
+                & (df["end"].le(earliest_end))
+                & (df["stream"].eq(stream))
+            ]
+        elif self.method.value == OperatorFilterMethod.After.value:
+            latest_start = max(op["end"], df.ts.min())
+            stream = op["stream"]
+            df_ops = df.loc[df["ts"].ge(latest_start) & df["stream"].eq(stream)]
+        elif self.method.value == OperatorFilterMethod.Before.value:
+            earliest_end = min(op["ts"], df["end"].max())
+            stream = op["stream"]
+            df_ops = df.loc[df["end"].le(earliest_end) & df["stream"].eq(stream)]
+        else:
```

Comment:
why is `OperatorFilterMethod=stack` not implemented?

Reply:
It has been implemented but we haven't imported in this PR.